### ISSUE TITLE
big.int.Mutable.popCount() has incorrect results when bit count crosses multiple of Limb

### ISSUE THREAD

**Issue by hryx** [ziglang/zig#13571]
### Zig Version

0.11.0-dev.202+a09a5ad57

### Steps to Reproduce and Observed Behavior

Because it's hard to do the math in the head, here is a program which will visually demonstrate the problem by printing the bits from bigints and the reported pop counts. Run with `zig run popcount.zig`.

Comment out the part that relies on the `Const.popCount()` implementation, or check out #13570 to try it out.

```zig
const std = @import("std");
const big = std.math.big;
const Managed = big.int.Managed;
const print = std.debug.print;
var gpa = std.heap.GeneralPurposeAllocator(.{}){};

pub fn main() !void {
    var a = try Managed.init(gpa.allocator());
    defer a.deinit();

    const limb_size = @bitSizeOf(big.Limb);
    try a.setString(2, "1" ** limb_size);
    a.negate();

    try printBits(&a, 18, limb_size * 2 - 2);
    try printBits(&a, 18, limb_size * 2 - 1);
    try printBits(&a, 18, limb_size * 2);
    try printBits(&a, 18, limb_size * 2 + 1);
    try printBits(&a, 18, limb_size * 2 + 2);
    try printBits(&a, 18, limb_size * 2 + 3);
    try printBits(&a, 18, limb_size * 2 + 4);
}

fn printBits(val: *const big.int.Managed, comptime buf_size: comptime_int, bit_count: usize) !void {
    var read: usize = 0;
    var buf: [buf_size]u8 = .{0} ** buf_size;
    val.toConst().writePackedTwosComplement(&buf, 0, bit_count, .Big);
    for (buf) |byte| {
        print("{b:0>8} ", .{byte});
        read += @popCount(byte);
    }

    var mut_res = try big.int.Managed.init(gpa.allocator());
    defer mut_res.deinit();
    try mut_res.popCount(val, bit_count);
    print(": read={} mut={}", .{ read, mut_res });

    // Comment this out if you don't have the Const.popCount impl
    const const_res = val.toConst().popCount(bit_count);
    print(" const={}", .{const_res});

    if (mut_res.toConst().orderAgainstScalar(read) != .eq) {
        print(" *", .{});
    }
    print("\n", .{});
}
```

Output below. Please scroll horizontally to see where which values have incorrect results. Apologies for the enormous numbers, I could only reproduce the issue when this particular value was interpreted with bit sizes that crossed 2x `@bitSizeOf(Limb)`.

If you haven't checked out my branch you will not see the `const=` results.

```
00000000 00000000 00111111 11111111 11111111 11111111 11111111 11111111 11111111 11111111 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000001 : read=63 mut=63 const=63
00000000 00000000 01111111 11111111 11111111 11111111 11111111 11111111 11111111 11111111 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000001 : read=64 mut=64 const=64
00000000 00000000 11111111 11111111 11111111 11111111 11111111 11111111 11111111 11111111 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000001 : read=65 mut=65 const=65
00000000 00000001 11111111 11111111 11111111 11111111 11111111 11111111 11111111 11111111 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000001 : read=66 mut=66 const=66
00000000 00000011 11111111 11111111 11111111 11111111 11111111 11111111 11111111 11111111 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000001 : read=67 mut=66 const=67 *
00000000 00000111 11111111 11111111 11111111 11111111 11111111 11111111 11111111 11111111 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000001 : read=68 mut=67 const=68 *
00000000 00001111 11111111 11111111 11111111 11111111 11111111 11111111 11111111 11111111 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000001 : read=69 mut=67 const=69 *
```

### Expected Behavior

`Mutable.popCount()` returns the same number of 1s as stored in the buffer and printed. Passing to grep verifies the discrepancy.

The new `Const` implementatin from #13570 does return results in accordance with the written bitstream.

**Comment by hryx** [comment]
I've narrowed it down a little. `bitNotWrap()` (called by `popCount()`) produces the wrong value when the most significant limb is all 1s, no need to negate. The value becomes wrong when `bit_count > @bitSizeOf(Limb) * 2 + 1` as reflected in the test program output above.

```zig
    // On 64-bit platforms. 129 is ok, 130 fails
    try a.set(maxInt(Limb));
    const original = try a.to(u130);
    try a.bitNotWrap(&a, .unsigned, 130);
    try testing.expect((try a.to(u130)) == ~original);
```

**Comment by matu3ba** [comment]
At least to me either testing.expect or the debug is broken. Or is this expected behavior:
```
(ins)(gdb) p/x tmp1 
$6 = 0xffffffffffffffff0000000000000000
(ins)(gdb) p/x tmp2 
$7 = 0xffffffffffffffff0000000000000000
(ins)(gdb) 
```
from
```zig
const std = @import("std");
const big = std.math.big;
const Managed = big.int.Managed;
const print = std.debug.print;
var gpa = std.heap.GeneralPurposeAllocator(.{}){};
pub fn main() !void {
    var a = try Managed.init(gpa.allocator());
    defer a.deinit();
    try a.set(std.math.maxInt(big.Limb));
    const original = try a.to(u130);
    try a.bitNotWrap(&a, .unsigned, 130);
    var tmp1 = try a.to(u130);
    var tmp2 = ~original;
    try std.testing.expect(tmp1 == tmp2);
}
```
(still failing with `error: TestUnexpectedResult`)

**Comment by matu3ba** [comment]
With Vexu's changes it looks like it is working:
```
(ins)(gdb) p/x tmp1
$2 = 0x1ffffffffffffffff0000000000000000
(ins)(gdb) p/x tmp2
$3 = 0x20dd03ffffffffffffffff0000000000000000
(ins)(gdb) p tmp1
$7 = 680564733841876926908302470789826871296
(ins)(gdb) p tmp2
$8 = 732877022673366638791900701566492939886002176
```

**Comment by matu3ba** [comment]
Using in `pub const Mutable = struct {`
```zig
    pub fn bitNotWrap(r: *Mutable, a: Const, signedness: Signedness, bit_count: usize) void {
        // TODO: This looks wrong.
        std.debug.print("1a:\n", .{});
        a.dump();
        r.copy(a.negate());
        std.debug.print("1a negated:\n", .{});
        r.dump();
        const negative_one = Const{ .limbs = &.{1}, .positive = false };
        std.debug.print("1negative_one:\n", .{});
        negative_one.dump();
        _ = r.addWrap(r.toConst(), negative_one, signedness, bit_count);
        std.debug.print("1addWrap result:\n", .{});
        r.dump();
    }
```
shows for  `u130` incorrectly:
```
1a:
ffffffffffffffff positive=true
1a negated:
ffffffffffffffff capacity=4 positive=false
1negative_one:
1 positive=false
1addWrap result:
0 ffffffffffffffff 1 capacity=4 positive=true
mutable after:
0 ffffffffffffffff 1 capacity=4 positive=true
mlen: 3
```
but for `u128`:
```
1a:
ffffffffffffffff positive=true
1a negated:
ffffffffffffffff capacity=4 positive=false
1negative_one:
1 positive=false
1addWrap result:
0 ffffffffffffffff capacity=4 positive=true
mutable after:
0 ffffffffffffffff capacity=4 positive=true
mlen: 2
```

Hence `m.bitNotWrap(a.toConst(), signedness, bit_count);` does the wrong thing for that input.

**Comment by matu3ba** [comment]
In `pub fn addWrap(r: *Mutable, a: Const, b: Const, signedness: Signedness, bit_count: usize) bool {`
`r.truncate(r.toConst(), signedness, bit_count);` has the wrong return value.

Interestingly enough, `~a = -a +1`, but Managed.bitWrap uses `-a + (-1)` to use the wraparound.

**Comment by matu3ba** [comment]
The cause of this error is uninitialized memory:
```zig
        if (r.addCarry(x, y)) {
            // There are two possibilities here:
            // - We overflowed req_limbs. In this case, the carry is ignored, as it would be removed by
            //   truncate anyway.
            // - a and b had less elements than req_limbs, and those were overflowed. This case needs to be handled.
            //   Note: after this we still might need to wrap.
            const msl = math.max(a.limbs.len, b.limbs.len);
            if (msl < req_limbs) {
                r.limbs[msl] = 1;
                r.len = req_limbs;
            } else {
                carry_truncated = true;
            }
        }
```
sets `r.len = req_limbs;`, so the upper limbs are not yet initialized.
Then in `r.truncate(r.toConst(), signedness, bit_count);` the zero-extension of the result fails, because `r.len` has already being updated:
```
            // Zero-extend the result
            if (req_limbs > r.len) {
                std.debug.assert(false);
                mem.set(Limb, r.limbs[r.len..req_limbs], 0);
            }
```
and we are using uninitialized memory.

Not sure why `const debug_safety = true;` has only effect with new compiler build and why the compiler panics with SIGSEV. I guess I'll rebuild LLVM with assertions.

**Comment by matu3ba** [comment]
@hryx Do you want to take a look? I'm feeling abit depressed, because I did not use `const debug_safety = true;`.

**Comment by hryx** [comment]
@matu3ba Nice catch! Zeroing the undefined limbs up to the new `r.len` does the trick. Both my small test case above and the program up in the original description now pass.

```diff
             const msl = math.max(a.limbs.len, b.limbs.len);
             if (msl < req_limbs) {
                 r.limbs[msl] = 1;
                 r.len = req_limbs;
+                mem.set(Limb, r.limbs[msl + 1 .. req_limbs], 0);
             } else {
                 carry_truncated = true;
             }
```

Does this seem like the correct solution to you?

> Not sure why const debug_safety = true; has only effect with new compiler build

Hm, actually I wasn't able to see any effect at all from setting this to `true`, at least not on this issue.

**Comment by matu3ba** [comment]
> Zeroing the extra limbs up to the new r.len does the trick.

Nice fix, it got very late and I tried other more bogous things. 

> Does this seem like the correct solution to you?

Are you sure `msl == req_limbs` is not possible? I need to have a look again on the `msl` computation.

If it also works with more than 2 limbs, expanding conversion from/to > 256 bit and different signs, then from a short look I suspect yes.
Last time I tested, the collapsing conversions were guarded by errors checking the necessary sizes (I think too conservative, but better safe than sorry.)

**Comment by hryx** [comment]
> Are you sure `msl == req_limbs` is not possible?

In the true branch where I added the line, `msl < req_limbs`. And in the else branch it shouldn't suffer the original problem because `r.len` isn't manually set without fixing the limb contents. So I think it's sound. I ran the behavior tests with the change compiled in and no problems came up.

> If it also works with more than 2 limbs, expanding conversion from/to > 256 bit and different signs, then from a short look I suspect yes.

Sounds good! Already tested crossing 4 limbs, but will check with negative values too. I might opt to make a test that checks pop count rather than e.g. comparing against a u258 so we don't have to use `SkipZigTest` for various backends.
